### DISCUSSION TITLE
NEC V20/V30 emulation

### DISCUSSION THREAD

**Discussion by Torinde** [dbalsom/marty#109] [Ideas]
NEC V20/V30 are kind of the most advanced pre-[80286](https://github.com/dbalsom/martypc/discussions/21) CPUs:
- used in some IBM PC/XT-compatibles
- the instructions from 80186 and real mode ones from 80286
- additional NEC instructions
- 8080 mode
- interesting FPU options (NEC μPD72091, μPD72191, D9008D) as shown at [2](https://www.cpushack.com/2021/09/01/necs-forgotten-fpus/),[3](https://www.cpu-world.com/forum/viewtopic.php?t=12863&view=previous),[1](https://www.vogons.org/viewtopic.php?t=68413) 
  - does MartyPC emulate 8087?
  - FPUs [8231/8232](https://en.wikipedia.org/wiki/Intel_8231/8232) are originally for 8080/8085/Z80, but were also mentioned as suitable for 8086 in its initial documentation.

There is also a variant with Z80 mode: NEC [μPD9002](https://retrocomputing.stackexchange.com/questions/26442/is-necs-pc-88va-compatible-with-n88-basic86) used in a PC-88 model.

Going into dual CPU options (most of them also not PC compatible):
- computers having x86 + Z80 [as two CPUs](https://en.wikipedia.org/wiki/Multiuser_DOS#CCP/M-86/80): [Rainbow 100](https://en.wikipedia.org/wiki/Rainbow_100), [Compupro System 816](https://sr.wikipedia.org/wiki/Compupro_System_816), [LSI Octopus](http://oldcomputer.info/8bit/hilger/index.htm), [LSI M4](https://forum.vcfed.org/index.php?threads/information-wanted-lsi-computers-m-four.1239479/), [CAL PC](https://nosher.net/archives/computers/pcn_1983-05-20_001), [North Star Advantage](https://en.wikipedia.org/wiki/North_Star_Computers#North_Star_Advantage_(1982)), most supporting some versions CP/M and DOS
 - 8085 and x86 - there were such computers having two CPUs with support for CP/M and DOS: [Zenith Z-100](https://en.wikipedia.org/wiki/Zenith_Z-100), [Compupro System 816](https://sr.wikipedia.org/wiki/Compupro_System_816)

Chips mentioned:
  - [ ] NEC V20 (80188/286+8080)
  - [ ] NEC V30 (80186/286+8080)
  - [ ] NEC V33 (80186/286+8080, 24-bit addresses - 286 bus?) 
  - [ ] 8087 FPU
  - [ ] μPD9002 (80186/286+Z80)
  - [ ] 8231 APU - z88dk-ticks has an implementation [here](https://github.com/z88dk/z88dk/blob/master/src/ticks/am9511.c).
  - [ ] 8232 FPU
  - [ ] μPD72091 FPU
  - [ ] μPD72191 FPU
  - [ ] D9008D FPU
  - [ ] NEC 72291 non-x87 FPU
  - [ ] 8085 - its undocumented instructions are implemented [here](https://github.com/z88dk/z88dk/blob/master/src/ticks/i8085_inst.c).
  - [ ] 8080
  - [x] 8088
  - [ ] 8086 - #108 
  - [ ] 80186
  - [ ] 80188
  - [ ] 8008 / CTC Datapoint 2200 - the origin of the microarchitecture utilized in the rest above

For me the most useful of those would be V20 and 8087 and the most interesting would be the NEC FPUs.

edit: added 8231/8085 links from [here](https://github.com/MockbaTheBorg/RunCPM/issues/192).

**Comment by dbalsom** [comment]
V20 is planned at some point, but I can't say it is a huge priority.  8087 would be nice but no immediate plans for it, because it is very complex.

8080/8085 coprocessors might be interesting to do. They are fairly simple devices in comparison to the 8088, and then I could add Space Invaders support to MartyPC :)

**Reply by Torinde** [reply]
That would be great!

For V20 - can you use code from [VirtualXT](https://github.com/andreas-jonsson/virtualxt/issues/18) (it seems they used similar hardware validation method as yourself) - or you plan to do it from scratch?

**Reply by dbalsom** [reply]
I am friends with Andreas and we exchange ideas (and I did get the hardware validation idea from him). But I would implement V20 myself. I would be generating a JSON test suite for it. VirtualXT also does not attempt to be cycle-accurate.  Without the V20 microcode this might be a challenge, but we know it must operate somewhat similarly.

**Comment by dbalsom** [comment]
I've started working on a framework to enable multiple CPU types, this has to be done before we can have anything other than an 8088

**Comment by dbalsom** [comment]
I've mostly completed research into the V20 CPU. 

I published a blog post, https://martypc.blogspot.com/2024/05/exploring-nec-v20-cpu.html

..and generated a test suite for the V20 cpu. https://github.com/singleStepTests/v20

Now I can use the test suite to validate MartyPC's V20 emulation.

**Comment by llyon1965** [comment]
Back in the day I replaced the 8088 with a V20 in my Commodore PC5, along with a  little tool that changed the RAM refresh rate I ended getting about 115% the speed of a bog standard IBM PC XT – it mattered!

**Comment by dbalsom** [comment]
MartyPC now has a functioning V20, with all native mode opcodes implemented except BRKEM (enter emulation mode). I will add 8088 emulation mode in the future. I am pondering ways to support emulation mode through Arduino8088 so that I can generate the relevant tests.

![image](https://github.com/dbalsom/martypc/assets/7229541/fc91d468-177a-4fcb-9e9b-12f634a2e11b)

When I say functioning, I mean it is basically an 8088 with relevant changes for V20 instruction differences, so that it passes my new V20 test suite in terms of register, memory and defined flag state. I still need to spend some time adding the V20's numerous optimizations to EA calculation, shifts/rotates, string operations, mul/div before it performs better than an 8088 in MartyPC.  I don't know if it will ever be cycle-perfect, but I think I can get it pretty close.

As to how you use it in MartyPC, I don't have base machine definitions with a V20, but what I am doing is maintaining a "compatible upgrade" list for each CPU type, and then you can specify `upgrade_type` via an `overlay.cpu`:

```toml
# Upgrade the CPU of the system to an NECV20. 
# Base machine type must have a compatible CPU (Intel 8088)
[[overlay]]
name = "cpu_v20"
    [overlay.cpu]
    upgrade_type = "NecV20"
```

The CPU overlay will eventually be a good place to add fields to support overclocking and other interesting features.

**Comment by dbalsom** [comment]
A functional, native-mode only V20 is now available in the latest release.
https://github.com/dbalsom/martypc/releases/tag/0.2.1
